---
title: "Parastichopus parvimenesis - Migrate-n"
output: 
  html_notebook:
      toc: true
---

This notebook will document a migrate-n analysis that I ran for *Parastichopus parvimensis*. I am working with Alison Haupt who received reviewer comments on a manuscript that urged her to do a migrate-n analysis.

First, the FASTA file taxon names are a bit funky, so I did some quick grepping in Textwrangler to distinguish taxon numbers from locality names:

Find: "_(\d\d)$"  
Replace: "-\\1"

Now taxon numbers are separated by a "-" while locality names can include "_"

```{r, message=F, echo=F}
library(stringr)
library(strataG)
library(phangorn)
library(phyclust)
```

## Read and check data
Read in the data and check alignment:
```{r}
ppar<-read.FASTA("Ppar.fasta")
image.DNAbin(ppar)
```


Pop off the population names, make a quick haplotype network. Looks pretty panmictic...

```{r}
pop<-gsub(names(ppar),pattern = "-\\d+",replacement="")

d <- dist.dna(ppar)
h <- pegas::haplotype(ppar)
h <- sort(h, what = "label")

net <- pegas::haploNet(h)
i<-stack(setNames(attr(h, "index"), rownames(h)))
i<-i[order(i$values),]
ind.hap<-table(hap=i$ind, pop=pop)

#play with scale.ratio to get appropriate branch lengths
plot(net,size=attr(net, "freq"), scale.ratio=10, pie=ind.hap,legend=F, labels=F,threshold=0, show.mutation=2)

plot.new()
legend("topleft", colnames(ind.hap), col=rainbow(ncol(ind.hap)), pch=19, ncol=2)

```

Wrote out the data as nexus format and edited it manually to migrate format.


## Run Modeltest
```{r}
ppar_phy<-phyDat(ppar)
#ppar_modeltest<-modelTest(ppar_phy,G=T, I=F)  - commented out because it takes awhile to run, so just load in the results.
ppar_modeltest<-load("modeltest_out.R")
ppar_modeltest<-ppar_modeltest[order(ppar_modeltest$BIC),]
ppar_modeltest
```

K80 is the best model, which is close enough to F84 and HKY that migrate uses. We will take the Gamma shape parameter from the HKY model

```{r}
# some funky code to get the gamma shape parameter
env<-attr(ppar_modeltest, "env")
HKY<-get("HKY+G", env)
eval(HKY, env=env)

```

The shape parameter is: 0.203

So the gamma distribution looks like this:

```{r}
plot(density(rgamma(1000,shape=0.10879)), xlim=c(0,0.5))
```
Using migrate-n to discretize that gamma into four categories we get:

rates=4: 0.333841 1.589418 3.974628 8.077175 
prob-rates=4: 0.561404 0.391364 0.046548 0.000684 

And the transition/transversion ratio is:

```{r}
TiTvRatio(ppar)
```
So we will use those numbers in our parmfiles

## Parmfile

### Migration Models

Symmetric gene flow
1. Panmixia

\*

2. Island Model

m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  
m	m	m	m	m	m	m	m	m	m	m	m	m	m	m  

3. Stepping-Stone

\*	s	0	0	0	0	0	0	0	0	0	0	0	0	0  
s	*	s	0	0	0	0	0	0	0	0	0	0	0	0  
0	s	*	s	0	0	0	0	0	0	0	0	0	0	0  
0	0	s	*	s	0	0	0	0	0	0	0	0	0	0  
0	0	0	s	*	s	0	0	0	0	0	0	0	0	0  
0	0	0	0	s	*	s	0	0	0	0	0	0	0	0  
0	0	0	0	0	s	*	s	0	0	0	0	0	0	0  
0	0	0	0	0	0	s	*	s	0	0	0	0	0	0  
0	0	0	0	0	0	0	s	*	s	0	0	0	0	0  
0	0	0	0	0	0	0	0	s	*	s	0	0	0	0  
0	0	0	0	0	0	0	0	0	s	*	s	0	0	0  
0	0	0	0	0	0	0	0	0	0	s	*	s	0	0  
0	0	0	0	0	0	0	0	0	0	0	s	*	s	0  
0	0	0	0	0	0	0	0	0	0	0	0	s	*	s  
0	0	0	0	0	0	0	0	0	0	0	0	0	s	*  

4. Stepping-stone -South dispersal

\*	0	0	0	0	0	0	0	0	0	0	0	0	0	0  
*	*	0	0	0	0	0	0	0	0	0	0	0	0	0  
0	*	*	0	0	0	0	0	0	0	0	0	0	0	0  
0	0	*	*	0	0	0	0	0	0	0	0	0	0	0  
0	0	0	*	*	0	0	0	0	0	0	0	0	0	0  
0	0	0	0	*	*	0	0	0	0	0	0	0	0	0  
0	0	0	0	0	*	*	0	0	0	0	0	0	0	0  
0	0	0	0	0	0	*	*	0	0	0	0	0	0	0  
0	0	0	0	0	0	0	*	*	0	0	0	0	0	0  
0	0	0	0	0	0	0	0	*	*	0	0	0	0	0  
0	0	0	0	0	0	0	0	0	*	*	0	0	0	0  
0	0	0	0	0	0	0	0	0	0	*	*	0	0	0  
0	0	0	0	0	0	0	0	0	0	0	*	*	0	0  
0	0	0	0	0	0	0	0	0	0	0	0	*	*	0  
0	0	0	0	0	0	0	0	0	0	0	0	0	*	* 

5. 2region1 - Two populations, Everything vs PQ+PCB

\* s  
s *  

5. 2region2 - Two populations, Everything vs. PCE, PQ, PCB

\* s  
s *  

### Parmfile
```{Bash}
################################################################################
# Parmfile for Migrate 3.6.4 [do not remove these first TWO lines]
menu=NO
nmlength=10
datatype=SequenceData
ttratio= 3.43
freqs-from-data=YES
seqerror-rate=0.0
categories=1
rates=4: 0.333841 1.589418 3.974628 8.077175 
prob-rates=4: 0.561404 0.391364 0.046548 0.000684 
autocorrelation=NO
weights=NO
interleaved=NO
fast-likelihood=NO
inheritance-scalars={1}
population-relabel={1 1 1 1 1 1 1 1 1 1 1 1 1 1 1}
usertree=RANDOMTREE
infile=../Ppar.mig
random-seed=AUTO
title=Parastichopus parvimensis
progress=YES
logfile=NO
print-data=NO
outfile=parastichopus_outfile.txt
pdf-outfile=parastichopus_outfile.pdf
use-M=YES
plot=NO
mathfile=mathfile
profile=ALL:QUICK
print-tree=NONE
write-summary=NO
aic-modeltest=NO
mig-histogram=NO
skyline=NO
theta=own:{0.01}
migration=own:{1000}
mutation=CONSTANT
fst-type=THETA
custom-migration={*}
geo=NO
bayes-update=YES
bayes-updatefreq=0.500000
bayes-posteriorbins=500 500
bayes-posteriormaxtype=ALL
bayes-file=YES:bayesfile
bayes-allfile=YES:1:bayesallfile
bayes-proposals= THETA METROPOLIS Sampler
bayes-proposals= MIG SLICE Sampler
bayes-priors= THETA WEXPPRIOR: 0.0 0.01 0.1000000 0.01000 
bayes-priors= MIG WEXPPRIOR: 0.000100 100000.000000 50000000.000000 100000.000000 
long-chains=1
long-inc=100
long-sample=75000
burn-in=20000
heating=YES:1:{1,1.5,3,100000}
heated-swap=YES
moving-steps=NO
long-chain-epsilon=INFINITY
gelman-convergence=No
replicate=YES:3
resistance=0.000100
end


```

## Harvest Marginal Likelihoods

#### A function for calculating model selection statistics
```{r}
bfcalcs<-function(df,ml="bezier.corrected"){
  df$thermodynamic<-as.numeric(df$thermodynamic)
  df$bezier.corrected<-as.numeric(df$bezier.corrected)
  df$harmonic<-as.numeric(df$harmonic)
    mlcol<-df[,ml] 
	bmvalue<-mlcol[which.max(mlcol)]
	lbf<-2*(mlcol-bmvalue)
	choice<-rank(-mlcol)
	modelprob<-exp(lbf/2)/sum(exp(lbf/2))
	dfall<-cbind(df,lbf,choice,modelprob)
	return(dfall)
}	
```


Here I am going through the three runs and harvesting out the likelihoods into a likelists object.

```{r harvest_likelihoods}
wd<-"?????"
setwd(wd)


for(r in 1:3){
  wd1<-paste(wd,"models",r,sep="")
  setwd(wd1)
  # start here if you just want to do one list
  likelist<-list() #initialize an empty list
  
    
    marglike<-data.frame(model=character(0),thermodynamic=numeric(0),bezier.corrected=numeric(0),harmonic.mean=numeric(0),stringsAsFactors=F) #initialize a data frame to take the values
    
    l=1 #initialize l
    
    for(i in c("panmixia","island","2region1","2region2","south_step","sym_step")){ #i<-"sym_step"
      wd3<-file.path(wd2,i)
      print(wd3)
      if(!file.exists(wd3)){next}
      setwd(wd3)
      outfile<-scan(file=paste(f,"outfile.txt",sep="_"),what="character",sep="\n") #scan in the outfile, separating at each newline
      

      
      #get the result from thermodynamic integration
      thermoline<-grep("(1a)",outfile,value=T) #find the line with the thermodynamic likelihood on it
      if(length(thermoline)==0){next}
      thermoline<-strsplit(thermoline,split="=",fixed=T) #split it up
      thermo<-as.numeric(substr(thermoline[[1]][2],start=1,stop=12)) #grab the thermodynamic likelihood
      bezistart<-grep("\\(",strsplit(thermoline[[1]][2],split="")[[1]])+1
      bezier<-as.numeric(substr(thermoline[[1]][2],start=bezistart,stop=bezistart+11)) #and the bezier-corrected value
      #get the harmonic mean
      harmoline<-grep("\\(2\\) H",outfile,value=T) #find the line with harmonic mean likelihood on it
      harmoline<-strsplit(harmoline,split="=",fixed=T) #grab the harmonic mean
      harmo<-as.numeric(substr(harmoline[[1]][2],start=1,stop=12))
      marglike[l,]<-c(i,thermo,bezier,harmo) #add this as a row to the data frame
      l=l+1
    }
    
    
    

  
# stop here if you just want one round  
  likelists[[r]]<-marglike
}

setwd(wd)
```

## Model Selection
On each round, and then bind them into a list
```{r model_selection}
modeltable1<-lapply(likelists[[6]],bfcalcs)
modeltable2<-lapply(likelists[[7]],bfcalcs)
modeltable3<-lapply(likelists[[8]],bfcalcs)
modeltables<-list("round1"=modeltable1,"round2"=modeltable2,"round3"=modeltable3)

head(modeltables[[1]],1)

```

## Plot marginal likelihood results from each of 3 runs
```{r}
#plot model selection graphics (blue dots with lines) - lines denote confidence intervals
m<-c("panmixia","hi_lo","1current","2currents","n.island","structure_panmixia","structure_currents","stepping.stone.1param","stepping.stone","empirical")
m<-c("panmixia","hi_lo","1current","2currents","n.island","stepping.stone.1param","stepping.stone","empirical")


pdf(file = "thermodynamic_marginal_likelihoods_final.pdf",width=8.5,height=3)
means<-list()
#plots<-list()
for(dataset in names(modeltables[[1]])){
  likes<-rbind(cbind(modeltables[[1]][[dataset]],rep=1), cbind(modeltables[[2]][[dataset]],rep=2), cbind(modeltables[[3]][[dataset]], rep=3))
  likes$model<-factor(likes$model, m)
  likes<-likes[!(is.na(likes$model)),]
  
  #likes<-likes[which(likes$bezier.corrected > max(likes$bezier.corrected)-100),]
  y.mean<-as.vector(by(likes$bezier.corrected,likes$model,mean))
  y.sd<-as.vector(by(likes$bezier.corrected,likes$model,sd))
  y.min<-y.mean-((y.sd/sqrt(3))*4.303)
  y.max<-y.mean+((y.sd/sqrt(3))*4.303)

    
  likes.mean<-data.frame(model=factor(m,m),y.mean,y.min,y.max,y.sd)
  means[[dataset]]<-likes.mean
  
  #l<-ggplot(data=likes, aes(x=model,y=bezier.corrected,colour=factor(rep), 
   #                                 shape=factor(rep), size=20 ))
  l<-ggplot(data=likes, aes(x=model,y=bezier.corrected))
  
  l<-l+geom_point(colour="blue", size=3)+
    geom_pointrange(data=likes.mean,y=y.mean,ymin=y.min,ymax=y.max, size=0.5)+
    scale_x_discrete(drop=FALSE)+
    theme(axis.text.y = element_text(size=16),legend.position="none",axis.title.x=element_text(size=16),axis.title.y=element_blank(),plot.title=element_text(size=20))+ggtitle(dataset)+ylab("Marginal Log-Likelihood")+
    coord_fixed(0.1)+ coord_flip()
  print(l)
#  plots<-c(plots,l)
}
dev.off()
```

# get rest of code from marginal_likelihood_final.Rmd